File: src/main/java/noh/jinil/boot/service/UserServiceImpl.java
package noh.jinil.boot.service;

import noh.jinil.boot.config.SecurityConfig;
import noh.jinil.boot.domain.dao.RoleRepository;
import noh.jinil.boot.domain.dao.UserRepository;
import noh.jinil.boot.domain.model.RoleEntity;
import noh.jinil.boot.domain.model.UserEntity;
import noh.jinil.boot.service.shared.RegisterUserInit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.annotation.Secured;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import javax.transaction.Transactional;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

@Service(value = UserServiceImpl.NAME)
public class UserServiceImpl implements UserService {

	public final static String NAME = "UserService";
	private final static Logger logger = LoggerFactory.getLogger(UserServiceImpl.class);

	@Autowired
	private UserRepository userRepository;

	@Autowired
	private RoleRepository roleRepository;

	@PostConstruct
	public void init() {
		userRepository.deleteAll();
		roleRepository.deleteAll();

		if (userRepository.count() == 0) {
			UserEntity adminEntity = new UserEntity();
			adminEntity.setUsername("1iTC5nOXFmeakKyM1NQ7EJr0TmB3");
			adminEntity.setPassword("admin");
			adminEntity.setEmail("for1self@gmail.com");

			adminEntity.setAuthorities(getAdminRoles());
			userRepository.save(adminEntity);

			UserEntity userEntity = new UserEntity();
			userEntity.setUsername("google_gabsikee@gmail.com");
			userEntity.setPassword("user");
			userEntity.setEmail("gabsikee@gmail.com");
			userEntity.setAuthorities(getUserRoles());

			userRepository.save(userEntity);
		}
	}

	@Override
	public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
		UserDetails userDetails = userRepository.findByUsername(username);
		if (userDetails == null)
			return null;

		/*
		Set<GrantedAuthority> grantedAuthorities = new HashSet<>();
		for (GrantedAuthority role : userDetails.getAuthorities()) {
			grantedAuthorities.add(new SimpleGrantedAuthority(role.getAuthority()));
		}
		*/

		return new org.springframework.security.core.userdetails.User(userDetails.getUsername(),
				userDetails.getPassword(), userDetails.getAuthorities());
	}

	@Override
	@Transactional
	@Secured(value = SecurityConfig.Roles.ROLE_ANONYMOUS)
	public UserEntity registerUser(RegisterUserInit init) {

		UserEntity userLoaded = userRepository.findByUsername(init.getUserName());

		if (userLoaded == null) {
			UserEntity userEntity = new UserEntity();
			userEntity.setUsername(init.getUserName());
			userEntity.setEmail(init.getEmail());

			userEntity.setAuthorities(getUserRoles());
			userEntity.setPassword(UUID.randomUUID().toString());
			userRepository.save(userEntity);
			logger.info("registerUser -> user created");
			return userEntity;
		} else {
			logger.info("registerUser -> user exists");
			return userLoaded;
		}
	}

	private List<RoleEntity> getAdminRoles() {
		return Collections.singletonList(getRole(SecurityConfig.Roles.ROLE_ADMIN));
	}

	private List<RoleEntity> getUserRoles() {
		return Collections.singletonList(getRole(SecurityConfig.Roles.ROLE_USER));
	}

	private RoleEntity getRole(String authority) {
		RoleEntity role = roleRepository.findByAuthority(authority);
		if (role == null) {
			return new RoleEntity(authority);
		} else {
			return role;
		}
	}

}

File: src/main/kotlin/noh/jinil/boot/service/impl/UserServiceImpl.kt
package noh.jinil.boot.service.impl

import com.google.firebase.auth.FirebaseToken
import noh.jinil.boot.config.SecurityConfig
import noh.jinil.boot.domain.entity.RoleEntity
import noh.jinil.boot.domain.entity.UserEntity
import noh.jinil.boot.domain.repository.RoleRepository
import noh.jinil.boot.domain.repository.UserRepository
import noh.jinil.boot.service.UserService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import java.util.*
import javax.annotation.PostConstruct

@Service(value = UserServiceImpl.NAME)
class UserServiceImpl : UserService {

    companion object {
        const val NAME = "UserService"
    }

    @Autowired
    private val userRepository: UserRepository? = null

    @Autowired
    private val roleRepository: RoleRepository? = null

    @PostConstruct
    fun init() {
        userRepository?.deleteAll()
        roleRepository?.deleteAll()
    }

    private fun getAdminRoles(): List<RoleEntity> {
        return Collections.singletonList(getRole(SecurityConfig.Roles.ROLE_ADMIN))
    }

    private fun getUserRoles(): List<RoleEntity> {
        return Collections.singletonList(getRole(SecurityConfig.Roles.ROLE_USER))
    }

    private fun getRole(authority: String): RoleEntity {
        return roleRepository?.findByAuthority(authority) ?: RoleEntity(authority)
    }

    //@Transactional
    //@Secured(value = [SecurityConfig.Roles.ROLE_ANONYMOUS])
    override fun registerUser(holder: FirebaseToken): UserEntity? {
        val userEntity = userRepository?.findByUsername(holder.uid)

        return if (userEntity == null) {
            logger.info("registerUser() email: ${holder.email}")
            val data = UserEntity().apply {
                username = holder.uid
                password = "password"
                email = holder.email
                level = 0
                name = holder.name
                type = "firebase"
                authorities = getUserRoles()
            }
            userRepository?.save(data)
        } else {
            null
        }
    }

    override fun verifyUser(holder: FirebaseToken): Boolean {
        logger.info("verifyUser() email: ${holder.email}")
        logger.debug("result: ${loadUserByUsername(holder.uid) != null}")
        return (loadUserByUsername(holder.uid) != null)
    }

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String?): UserDetails? {
        logger.info("loadUserByUsername() id: $username")
        if (username == null) {
            return null
        }

        try {
            val userEntity = userRepository?.findByUsername(username) ?: return null
            logger.debug("->userEntity: $userEntity")
            return User(userEntity.username, userEntity.password, userEntity.authorities)
        } catch (e: EmptyResultDataAccessException) {
            e.printStackTrace()
            return null
        }

        /*
		Set<GrantedAuthority> grantedAuthorities = new HashSet<>();
		for (GrantedAuthority role : userDetails.getAuthorities()) {
			grantedAuthorities.add(new SimpleGrantedAuthority(role.getAuthority()));
		}
		*/
    }
}

private val logger = LoggerFactory.getLogger(UserServiceImpl::class.java)